Add a best practice for not using internal APIs (#32759)
gradle/gradle: platforms/documentation/docs/src/docs/userguide/best-practices/bp_general.adoc

[[best_practices_general]]
= General Gradle Best Practices

== Do Not Use Internal APIs

> Do not use APIs from a package where any segment of the package is `internal`, or types that have `Internal` or `Impl` as a suffix in the name.

Reviewer: It might be a good idea to point to everything that the Gradle team thinks is public from here? I know we have https://github.com/gradle/gradle/blob/9260f503383753c0de270f3d1843b8ad328825dd/build-logic-commons/basics/src/main/kotlin/gradlebuild/basics/PublicApi.kt#L12-L50, though I am not sure there is some user facing documentation about this.

Author: I'm happy to add that a follow once once such page exists, unless you are saying i should that copy that list here.

Reviewer: I suppose I'd like to get some input from @lkasso here...

Author: > Linking to a published public API artifact containing only the public API classes might be more useful.

I was under the impression that Gradle has no such artifact. AGP and KGP does have those so I could link to those. WDYT?

Reviewer 2: It will be documented in: https://github.com/gradle/gradle/pull/32902. @liutikas, feel free to point to it.

Author: Let's do that as a follow up.

=== Explanation

Using internal APIs is inherently risky and can cause significant problems during upgrades.

> Gradle and many plugins (such as Android Gradle Plugin and Kotlin Gradle Plugin) treat internal APIs, such as those within `internal` packages, as subject to unannounced breaking changes during any new Gradle release, even during minor releases.

Author: Done

There have been numerous cases where even highly experienced plugin developers have been bitten by usage of such APIs leading to breakages for their users.

> If you require a specific functionality that is missing, it’s best to submit a feature request.

Author: Done

[[best_practices_general]]
= General Gradle Best Practices

== Do Not Use Internal APIs

> Do not use APIs that are inside an `internal` Java package or contain `Internal` or `Impl`

Reviewer 3: 💅 I think we should be more precise about defining an `internal` package as any package where any of the segments in the package name is `internal`.

Also we should say not just "contains" but has Internal/Impl as a suffix to the name, as there might be edge cases here where this includes valid public types.

Author: Used [lkasso](https://github.com/lkasso)'s suggestion for this. Let me know if the updated version is better.

> Do not use APIs that are inside an `internal` Java package or contain `Internal` or `Impl`

Reviewer 2: A possible update:

```suggestion
Do not use APIs from a package where any segment of the package name is `internal`, or types that have `Internal` or `Impl` as a suffix in their name.
```

in the name of the type.

=== Explanation

> Using internal APIs is inherently risky and can cause significant problems during upgrades.

Reviewer 3: 💅 Formatting note: Please start each sentence on a new line in a `.md` or `.adoc` file.  Separate paragraphs with a blank line.

> Gradle and many plugins (such as Android Gradle Plugin and Kotlin Gradle Plugin) treat internal

Reviewer 3: 🤔 Should we call out that sometimes plugins like AGP and KGP have used internals and been bitten by it, and don't copy them?

> APIs, such as those within internal packages, as subject to breaking changes, even during minor

Reviewer 3: ```suggestion
APIs, such as those within internal packages, as subject to unannounced breaking changes during any new Gradle release, even during minor
```

I think we need to really hammer the point that there are no backwards or forwards compatibility guarantees here.

releases.

If you require a specific functionality that is missing, it’s best to submit a feature request.

> As a temporary workaround consider copying the necessary code into your own codebase.

Reviewer 3: ```suggestion
As a temporary workaround consider copying the necessary code into your own codebase and extending a Gradle public type with your own custom implementation using the copied code.
```

I think just saying copying isn't really enough guidance.

=== Example

==== Don't Do This

====
include::sample[dir="snippets/bestPractices/avoidInternal/kotlin",files="build.gradle.kts[tags=avoid-this]"]
include::sample[dir="snippets/bestPractices/avoidInternal/groovy",files="build.gradle[tags=avoid-this]"]
====

==== Do This Instead

====
include::sample[dir="snippets/bestPractices/avoidInternal/kotlin",files="build.gradle.kts[tags=do-this]"]
include::sample[dir="snippets/bestPractices/avoidInternal/groovy",files="build.gradle[tags=do-this]"]
====

=== Tags

> `#upgrades`

Reviewer 3: Can you add an entry in the `tags_reference.adoc` for this as well?
